### ISSUE TITLE
Notifications do not get fired

### ISSUE THREAD

**Issue by dennisgec** [realm/realm-cocoa#4842]
## Goals
I would like to receive real time updates on my post list in direct conversations.

## Expected Results
The notification callback should be fired every time the a new post is added to the list of posts in the direct object.

## Actual Results
Unexpected behavior by the callback. Even though testing and monitoring the database within realm browser, the callback for new posts, gets fired periodically and not every time.

### Observe Method
```swift
/**
     * Observes all posts from a direct conversation.
     *
     * - parameter direct: The direct that should be observed its posts.
     * - parameter update: An update callback containing an updated aaray of posts for that specific direct.
     */
    public func observePosts(
        fromDirect direct: Direct,
        update: @escaping (_ posts: [Post]) -> Void) {
        
        observerTokens.append(direct
            .posts
            .addNotificationBlock() {
            
            changes in
            
            switch changes {
                
            case .error(_):
                
                break
                
            default:
                
                update(Array(direct.posts))
            }
        })
    }
```

### Direct Model

```swift
/**
 * Direct.swift
 * Model
 *
 * Created by Dennis Gecaj on 9/28/16.
 *
 * The model of the direct.
 *
 * - author: Dennis Gecaj
 * - copyright: © 2016 Yello Inc. All rights reserved.
 * - version: 0.1
 */

import RealmSwift

open class Direct: Object {
    
    //  MARK: - PROPERTIES
    
    //  MARK: - Data
    
    /// A date indicating when the direct has been last updated.
    public dynamic var updatedAt: Date = Date()
    
    /// The members of the direct conversation.
    public var members = List<Profile>()
    /// The posts of the direct conversation.
    public var posts = List<Post>()
    
    /// The string id of the direct.
    public dynamic var id = String()
    
    //  MARK: - Static
    
    internal struct PropertyKeys {
        
        static let id = "id"
    }
    
    //  MARK: - METHODS
    
    //  MARK: - Initializer
    
    /**
     * Initializes and returns a newly allocated direct object.
     *
     * - parameter members: The members of the direct conversation.
     * - parameter posts: The posts of the direct conversation.
     * - parameter id: The string id of the direct.
     */
    public convenience init(
        members: List<Profile>?,
        posts: List<Post>?,
        id: String) {
        
        self.init()
        
        if let members = members {
            
            self.members = members
        }
        if let posts = posts {
            
            self.posts = posts
        }
        
        self.id = id
    }
    
    //  MARK: - Realm
    
    override open static func primaryKey() -> String? {
        
        return PropertyKeys.id
    }
}
```

### Post Model
```swift
/**
 * Post.swift
 * Model
 *
 * Created by Dennis Gecaj on 8/29/16.
 *
 * The model of the post.
 *
 * - author: Dennis Gecaj
 * - copyright: © 2016 Yello Inc. All rights reserved.
 * - version: 0.1
 */

import RealmSwift

open class Post: Object {
    
    //  MARK: - PROPERTIES
    
    //  MARK: - Data
    
    /// An integer as the type of the post.
    public dynamic var type: String = ContentType.text.rawValue
    /// The content of the post.
    public dynamic var content: AnyContent!
    /// An optional location of the post.
    public dynamic var location: Location?
    /// A boolean indicating whether the post is kept.
    public dynamic var keep: Bool = false
    /// A boolean indicating whether the post is currently uploading.
    public dynamic var isUploading: Bool = false
    /// The profile of the creator, the user who posted.
    public dynamic var profile: Profile?
    /// A date indicating when the post was created.
    public dynamic var createdAt: Date!
    
    /// A list of looks of the post.
    public var looks = List<Look>()
    /// A list of likes of the post.
    public var likes = List<Like>()
    
    /// The string id of the post.
    public dynamic var id = String()
    
    //  MARK: - Static
    
    public enum ContentType: String {
        
        case text = "text"
        case image = "image"
        case location = "location"
        case weather = "weather"
        case voice = "voice"
        case story = "story"
    }
    
    internal struct PropertyKeys {
        
        static let id = "id"
    }
    
    //  MARK: - METHODS
    
    //  MARK: - Initializer
    
    /**
     * Initializes and returns a newly allocated post object.
     *
     * - parameter type: An integer as the type of the post.
     * - parameter content: The content of the post.
     * - parameter location: An optional location of the post.
     * - parameter weather: An optional weather info of the post.
     * - parameter keep: A boolean indicating whether the post is kept.
     * - parameter profile: The profile of the user who posted.
     * - parameter createdAt: A date indicating when the post was created.
     * - parameter id: The string id of the post.
     */
    public convenience init(
        type: String,
        content: AnyContent,
        location: Location?,
        weather: Weather?,
        keep: Bool,
        profile: Profile?,
        createdAt: Date,
        id: String) {
        
        self.init()
        
        self.type = type
        self.content = content
        self.location = location
        self.keep = keep
        self.profile = profile
        self.createdAt = createdAt
        
        self.id = id
    }
    
    //  MARK: - Realm
    
    override open static func primaryKey() -> String? {
        
        return PropertyKeys.id
    }
}
```

## Version of Realm and Tooling
Realm framework version: 2.5.1

Xcode version: 8.3

iOS/OSX version: 10.3

**Comment by istx25** [comment]
Hi @dennisgec. Thanks for reaching out. I wanted to let you know that we've received your issue and that someone will review what you've shared and follow-up soon.

**Comment by dennisgec** [comment]
@jpsim I looked over the code again, could it be because I don't properly indicate the inverse relationship from the post back to the profile that I see the unexpected behavior?

**Comment by TimOliver** [comment]
Hi @dennisgec!

Hmm, I tried reviewing your code, but I think we need some more information to try and work out what's going on here. Can you please explain what sort of transaction you're performing when the expected notifications aren't occurring? Additionally, are you doing any of this on different threads?

I'm not sure if it would help, but could you post the model definition for `Profile` as well? If all else fails, would you be willing to share your project with us?

Since you're adding a notification straight to the `List` property of an object, you should reliably get a notification each time something causes that `List` property to mutate.

As an aside, notifications aren't triggered for computed properties like `LinkingObject` (we're looking into that one), but as far as I can gather, that's not being used at all here.

Thanks!

**Comment by Taco55** [comment]
This looks like the same issue as I posted at #4800. Could it be?

**Comment by TimOliver** [comment]
@Taco55 Interesting point!

@dennisgec Are you using the Realm Mobile Platform at all? Or is this a completely offline Realm?

**Comment by dennisgec** [comment]
@TimOliver I actually experienced this issue only using the offline Realm.

**Comment by TimOliver** [comment]
@dennisgec Hmm fair enough then. That being said, the notifications system is an inherently local operation, so they could still be related.

If that's the case, we pushed a new fix to `master` yesterday that fixed the problem in #4800. Could you please try integrating it into your project as well and seeing if they are related?

```
pod 'Realm', git: 'https://github.com/realm/realm-cocoa.git', branch: 'master', submodules: true
pod 'RealmSwift', git: 'https://github.com/realm/realm-cocoa.git', branch: 'master', submodules: true
```

**Comment by TimOliver** [comment]
We haven't heard from @dennisgec in a while. Hopefully that means the new branch solved the issue for him. :)

I'm going to close this issue off. Dennis, if you're still having trouble, please don't hesitate to reopen and follow up. :)
